### PR TITLE
SOF HDA/HDMI enabling -- Part1

### PR DESCRIPTION
**Pull request by keyonjie**
This part is the major part:
1. enable HDA bus initialization and codec probing.
2. use hda_ext_stream and expand SOF stream management to support HDA stream.
3. make sure it doesn't break the existed DSP non-HDA mode.

The later part(will send PR after this is merged) will include:
1. DPIB position update.
2. Legacy HDA/HDMI mode support.

### CODE REVIEW

**Comment by keyonjie** [comment]
@lgirdwood can you help review this?

File: sound/soc/sof/sof-acpi-dev.c [hunk 1]
```diff
@@ -149,13 +149,8 @@ static void sof_acpi_fw_cb(const struct firmware *fw, void *context)
...
 
 	/* register PCM and DAI driver */
-	priv->pdev_pcm =
-		platform_device_register_data(dev, "sof-audio", -1,
-					      sof_pdata, sizeof(*sof_pdata));
-	if (IS_ERR(priv->pdev_pcm)) {
-		dev_err(dev, "Cannot register device sof-audio. Error %d\n",
-			(int)PTR_ERR(priv->pdev_pcm));
-	}
+	sof_create_platform_device(priv);
```
Comment (lgirdwood) [inline sound/soc/sof/sof-acpi-dev.c]: Need to check return value.

File: sound/soc/sof/intel/hda-ctrl.c [hunk 2]
```diff
@@ -131,3 +131,27 @@ int hda_dsp_ctrl_get_caps(struct snd_sof_dev *sdev)
 	return ret;
 }
 
+void hda_dsp_ctrl_enable_miscbdcge(struct snd_sof_dev *sdev, bool enable)
```
Comment (lgirdwood) [inline sound/soc/sof/intel/hda-ctrl.c]: static inline

File: sound/soc/sof/intel/hda-bus.c [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+			continue;
+		drv = drv_to_hdac_driver(codec->dev.driver);
+		if (drv->unsol_event)
+			drv->unsol_event(codec, res);
+	}
+}
+
+#endif
+
+static void sof_hda_writel(u32 value, u32 __iomem *addr)
```
Comment (lgirdwood) [inline sound/soc/sof/intel/hda-bus.c]: Do we really need these wrappers ?

Comment (keyonjie) [reply sound/soc/sof/intel/hda-bus.c]: as I commented before, yes, as the writel()s are only macros, we need function wraps and assigning them to ops.

File: sound/soc/sof/sof-acpi-dev.c (hunk 1 above)
Comment (keyonjie) [reply sound/soc/sof/sof-acpi-dev.c]: this callback function actually will be removed in the next commit.

Comment (keyonjie) [reply sound/soc/sof/sof-acpi-dev.c]: but let me remove this change here in the update versions.

Comment (lgirdwood) [reply sound/soc/sof/sof-acpi-dev.c]: @keyonjie feel free to squash these patches per file if it makes things easier. They will be upstreamed as a single patch per file.

**Comment by keyonjie** [comment]
On 2018年08月31日 16:36, Liam Girdwood wrote:
> lgirdwood commented on this pull request.
> 
> 
> 
>> @@ -149,13 +149,8 @@ static void sof_acpi_fw_cb(const struct firmware *fw, void *context)
>   	}
>   
>   	/* register PCM and DAI driver */
> -	priv->pdev_pcm =
> -		platform_device_register_data(dev, "sof-audio", -1,
> -					      sof_pdata, sizeof(*sof_pdata));
> -	if (IS_ERR(priv->pdev_pcm)) {
> -		dev_err(dev, "Cannot register device sof-audio. Error %d\n",
> -			(int)PTR_ERR(priv->pdev_pcm));
> -	}
> +	sof_create_platform_device(priv);
> 
> @keyonjie feel free to squash these patches per file if it makes things easier. They will be upstreamed as a single patch per file.

OK, good point, I thought about that but didn't do in that way yet. Thanks.

>

**Comment by keyonjie** [comment]
@lgirdwood it is updated.
what I have minor concern is the commit: 
ALSA: HDA: export process_unsol_events() which can be used without hdac_bus library

I discussed with Mengdong about it, we think better to export this library function and reuse it in SOF.

**Pull request merged**